### PR TITLE
Polarimetry: static stream + testcase

### PR DESCRIPTION
**Pull request by Sabrina85**

### CODE REVIEW

File: src/odemis/acq/stream/_static.py [hunk 1]
```diff
@@ -256,52 +256,55 @@ def __init__(self, name, data, *args, **kwargs):
         data = [d.getData() if isinstance(d, model.DataArrayShadow) else d for d in data]
 
         # find positions of each acquisition
-        # tuple of 2 floats -> DataArray: position on SEM -> data
-        self._sempos = {}
+        # tuple of 2 floats + 1 string) -> DataArray: position on SEM + position of polarization analyzer -> data
+        self._pos = {}
         for d in data:
+            if not model.MD_POL_MODE:
```
Comment (philipwinklerdelmic) [inline src/odemis/acq/stream/_static.py]: shouldn't that be something like "if MD_POL_MODE not in d.metadata"?

File: src/odemis/acq/stream/_static.py [hunk 2]
```diff
@@ -390,7 +393,7 @@ def _setBackground(self, data):
 
         # check it's compatible with the data
         data = img.ensure2DImage(data)
-        arpole = data.metadata[model.MD_AR_POLE] # we expect the data has AR_POLE
+        arpole = data.metadata[model.MD_AR_POLE]  # we expect the data has AR_POLE
```
Comment (philipwinklerdelmic) [inline src/odemis/acq/stream/_static.py]: maybe you can also import MD_AR_POLE in the beginning? Otherwise, you sometimes have to use the model.MD...
prefix, and sometimes not, which might be confusing

File: src/odemis/acq/stream/_static.py [hunk 3]
```diff
@@ -256,52 +256,55 @@ def __init__(self, name, data, *args, **kwargs):
...
         # Cached conversion of the CCD image to polar representation
         # TODO: automatically fill it in a background thread
-        self._polar = {} # dict tuple 2 floats -> DataArray
+        self._polar = {}  # dict tuple 2 floats -> DataArray
 
-        # SEM position displayed, (None, None) == no point selected
-        self.point = model.VAEnumerated((None, None),
-                     choices=frozenset([(None, None)] + list(self._sempos.keys())))
+        # SEM position displayed, (None, None, None) == no point selected (x, y, pol)
+        self.point = model.VAEnumerated((None, None, None),
```
Comment (pieleric) [inline src/odemis/acq/stream/_static.py]: I don't think it's a good idea to just extend `.point` to a 3rd dimension. First, that would require changes to "all" code using this VA (admittedly, it's not that many, probably just a few places in the GUI). Second, because SEM pos and polarization are unrelated, and so will not likely change simultaneously. The GUI controls would either change the X/Y or the polarization.
=> leave `.point` and create `.polarization`.

Also, add a check that every position has the same set of polarizations. Probably, you should also not add this VA if the only polarization is `None`.

File: src/odemis/acq/stream/_static.py (hunk 1 above)
Comment (pieleric) [reply src/odemis/acq/stream/_static.py]: Instead of checking for the metadata, just read it this way:
`self._pos[d.metadata[MD_POS] + (d.metadata.get(MD_POL_MODE, None),)] = ...`
or to be more readable:
```
sempos = d.metadata[MD_POS]
pol = d.metadata.get(MD_POL_MODE, None)
self._pos[sempos + (pol,)] = ...
```

File: src/odemis/acq/stream/_static.py [hunk 4]
```diff
@@ -364,7 +367,7 @@ def _updateImage(self):
 
         pos = self.point.value
         try:
-            if pos == (None, None):
+            if pos == (None, None, None):
```
Comment (pieleric) [inline src/odemis/acq/stream/_static.py]: Just check for `(None, None)`. (no need to check for the polarization)

File: src/odemis/acq/stream/_static.py [hunk 5]
```diff
@@ -354,6 +369,30 @@ def _project2Polar(self, pos):
...
+        """
+        bg_data = self.background.value  # list containing bg images
+        if bg_data is None:
+            return None
+        if pol_mode:
+            for bg_im in bg_data:
+                if bg_im.metadata[MD_POL_MODE] == data.metadata[MD_POL_MODE]:
+                    # should be only one bg image with the same metadata entry
+                    bg = bg_im
+                    break
```
Comment (pieleric) [inline src/odemis/acq/stream/_static.py]: directly `return bg_im`

File: src/odemis/acq/stream/_static.py [hunk 6]
```diff
@@ -354,6 +369,30 @@ def _project2Polar(self, pos):
...
+            else:
+                return None
+                logging("No AR background with polarization ....")
```
Comment (pieleric) [inline src/odemis/acq/stream/_static.py]: Put the logging _before_ the return, and replace the `...` with the pol_mode.

File: src/odemis/acq/stream/_static.py [hunk 7]
```diff
@@ -385,47 +428,87 @@ def _updateImage(self):
     def _onPoint(self, pos):
         self._shouldUpdateImage()
 
+    def _onPolarization(self, pos):
+        self._shouldUpdateImage()
+
     def _setBackground(self, data):
         """Called when the background is about to be changed"""
```
Comment (pieleric) [inline src/odemis/acq/stream/_static.py]: Please indicate the type of `data` (None, DataArray, or list of DataArrays), and add a not that it only returns either None or a list of DataArrays.

File: src/odemis/acq/stream/_static.py [hunk 8]
```diff
@@ -385,47 +428,87 @@ def _updateImage(self):
...
-        # check the AR pole is at the same position
-        for r in self.raw:
-            if r.metadata[model.MD_AR_POLE] != arpole:
-                logging.warning("Pole position of background data %s is "
-                                "different from the data %s.",
-                                arpole, r.metadata[model.MD_AR_POLE])
+            return data
 
-        return data
+        # data = img.ensure2DImage(data)
```
Comment (pieleric) [inline src/odemis/acq/stream/_static.py]: Let's drop all this commented duplicateblock?

File: src/odemis/acq/test/stream_test.py [hunk 9]
```diff
@@ -2822,7 +2822,7 @@ def on_im(im):
 
         e.clear()
         ars.background.value = calib
-        numpy.testing.assert_equal(ars.background.value, calib[0, 0, 0])
+        numpy.testing.assert_equal(ars.background.value[0], calib[0, 0, 0])
```
Comment (pieleric) [inline src/odemis/acq/test/stream_test.py]: This change highlights the issue of automatically converting a single DataArray with a list. It's quite confusing for the (code) user/reader. You write something and next time you read it, it's inside a list?!

I think this is a good reason to reconsider the "trick" to make the background data always a list. I'd suggest that it accepts both a None, DataArray, or list of DataArray, and keeps the value as set. The trick is only to slightly simplify the code, but if that makes the rest of the code more complicated, that's not useful.

Comment (Sabrina85) [reply src/odemis/acq/test/stream_test.py]: I changed it now so None, DataArray, or list of DataArray but I think it is not the most elegant solution. Also I am not sure what you mean with using the value as a set and where...Would be happy to have some more explanation on that :)

File: src/odemis/acq/stream/_static.py [hunk 10]
```diff
@@ -431,16 +426,23 @@ def _onPoint(self, pos):
...
         self._shouldUpdateImage()
 
-    def _setBackground(self, data):
-        """Called when the background is about to be changed"""
-        if data is None:
+    def _setBackground(self, bg_data):
+        """
+        Called when the background is about to be changed
+        :param bg_data: (None, DataArray or list of DataArrays) background image(s)
+        :return: (None, DataArray or list of DataArrays)
```
Comment (pieleric) [inline src/odemis/acq/stream/_static.py]: Yes, that's helps to have a full docstring. I'd recommend to add _even_ more: you can add a `raises` after the `return` to indicate which types of Exceptions are raised when. Here, you'd say something like `ValueError`: the background data is not compatible with the data (ex: not the same shape or data type or polarization).

File: src/odemis/acq/stream/_static.py [hunk 11]
```diff
@@ -346,7 +346,7 @@ def _project2Polar(self, pos):
                 # TODO: could use the size of the canvas that will display
                 # the image to save some computation time.
 
-                # get bg image if existing, else "pass-through"
+                # get bg image if existing, else "pass-through" = MD_POL_NONE
```
Comment (pieleric) [inline src/odemis/acq/stream/_static.py]: That comment is a little confusing. I'd reword it this way: `Get bg image, if existing. It must match the polarization (defaulting to MD_POL_NONE).`

File: src/odemis/acq/stream/_helper.py [hunk 12]
```diff
@@ -36,6 +36,7 @@
 import time
 
 from ._base import Stream, UNDEFINED_ROI, POL_6POS, POL_POSITIONS
+from odemis.model._metadata import MD_POL_NONE
```
Comment (pieleric) [inline src/odemis/acq/stream/_helper.py]: Please import from `odemis.model`. (no need for `_metadata`)
